storage/mysql/admin_storage.go

```diff
@@ -0,0 +1,394 @@
+)
+
+// duplicatePolicyMap maps storage enums to trillian.DuplicatePolicy enums,
+// which differ slightly.
+var duplicatePolicyMap = map[string]trillian.DuplicatePolicy{
+	"NOT_ALLOWED": trillian.DuplicatePolicy_DUPLICATES_NOT_ALLOWED,
+	"ALLOWED":     trillian.DuplicatePolicy_DUPLICATES_ALLOWED,
+}
+
+// NewAdminStorage returns a MySQL storage.AdminStorage implementation backed by DB.
+func NewAdminStorage(db *sql.DB) storage.AdminStorage {
+	return &mysqlAdminStorage{db}
+}
+
+// mysqlAdminStorage implements storage.AdminStorage
+type mysqlAdminStorage struct {
+	db *sql.DB
+}
+
+func (s *mysqlAdminStorage) Snapshot(ctx context.Context) (storage.ReadOnlyAdminTX, error) {
+	return s.Begin(ctx)
+}
+
+func (s *mysqlAdminStorage) Begin(ctx context.Context) (storage.AdminTX, error) {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+	return &adminTX{tx: tx}, nil
+}
+
+type adminTX struct {
+	tx *sql.Tx
+
+	// mu guards *direct* reads/writes on closed, which happen only on
+	// Commit/Rollback/IsClosed/Close methods.
+	// We don't check closed on *all* methods (apart from the ones above),
+	// as we trust tx to keep tabs on its state (and consequently fail to do
+	// queries after closed).
+	mu     sync.RWMutex
+	closed bool
+}
+
+func (t *adminTX) Commit() error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.closed = true
+	return t.tx.Commit()
+}
+
+func (t *adminTX) Rollback() error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.closed = true
+	return t.tx.Rollback()
+}
+
+func (t *adminTX) IsClosed() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.closed
+}
+
+func (t *adminTX) Close() error {
+	// Acquire and release read lock manually, without defer, as if the txn
+	// is not closed Rollback() will attempt to acquire the rw lock.
+	t.mu.RLock()
+	closed := t.closed
+	t.mu.RUnlock()
+	if !closed {
+		return t.Rollback()
+	}
+	return nil
+}
+
+func (t *adminTX) GetTree(ctx context.Context, treeID int64) (*trillian.Tree, error) {
+	stmt, err := t.tx.Prepare(selectTreeByID)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+	return readTree(stmt.QueryRow(treeID))
+}
+
+// There's no common interface between sql.Row and sql.Rows(!), so we have to
+// define one.
+type row interface {
+	Scan(dest ...interface{}) error
+}
+
+func readTree(row row) (*trillian.Tree, error) {
+	tree := &trillian.Tree{}
+
+	// Enums and Datetimes need an extra conversion step
+	var treeState string
+	var treeType string
+	var hashStrategy string
+	var hashAlgorithm string
+	var signatureAlgorithm string
+	var duplicatePolicy string
+	var createDatetime string
+	var updateDatetime string
+
+	err := row.Scan(
+		&tree.TreeId,
+		&treeState,
+		&treeType,
+		&hashStrategy,
+		&hashAlgorithm,
+		&signatureAlgorithm,
+		&duplicatePolicy,
+		&tree.DisplayName,
+		&tree.Description,
+		&createDatetime,
+		&updateDatetime,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert all things!
+	if ts, ok := trillian.TreeState_value[treeState]; ok {
```

Comment:
gocyclo isn't complaining. I'd rather keep it here, as otherwise functions will multiply really quickly.

I tried a bunch of alternative ways of abstracting/simplifying this, but I couldn't find a way around the barrage of if/else/return err that Golang likes so much.